resolver/npm.go

```diff
@@ -0,0 +1,78 @@
+package resolver
+
+import (
+	"encoding/json"
+	"os"
+	"strings"
+)
+
+func ProjectsFromNpmLockfileV3(filename string) ([]Project, error) {
+	byteValue, err := os.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	var packageLock PackageLockV3
+
+	json.Unmarshal(byteValue, &packageLock)
+
+	return asNpmProjects(packageLock), nil
+}
+
+type NpmProject struct {
+	name     string
+	version  string
+	optional bool
+}
+
+var _ Project = (*NpmProject)(nil)
+
+func (p NpmProject) Name() string {
+	return p.name
+}
+
+func (p NpmProject) Optional() bool {
+	return p.optional
+}
+
+func (p NpmProject) Version() string {
+	return p.version
+}
+
+type PackageLockV3 struct {
+	Name     string                `json:"name"`
+	Version  string                `json:"version"`
+	Packages map[string]NpmPackage `json:"packages"`
+}
+
+type NpmPackage struct {
+	Name    string `json:"name"`
+	Version string `json:"version"`
+}
+
+func asNpmProjects(packageLock PackageLockV3) []Project {
+	projectSet := make(map[string]NpmProject, len(packageLock.Packages)-1)
+
+	for pkg, entry := range packageLock.Packages {
+		if pkg == "" {
+			// skip the top-level package as it refers to the project itself
+			continue
+		}
+
+		projectSet[pkg] = NpmProject{
+			// Remove the `node_modules/` root directory prefix from each `pkg`
+			name:    strings.Replace(pkg, "node_modules/", "", 1),
+			version: entry.Version,
+			// optional packages that are included as dependencies in the build will be declared at the top
+			// level of packageLock.Packages, so we can explicitly mark optional as false here
+			optional: false,
```

Comment:
When compared to the yarn.lock equivalent, determining if an `optional` package is used is simple. If the package is _actually_ used in the build, it will have an entry at the top level of the `package-lock.json` file.

Since `ossls` omits optional packages from the output, we can set this value as `false` directly to include all listed packages.